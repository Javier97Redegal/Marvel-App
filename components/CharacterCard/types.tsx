type CharacterCardProps = {
    character: {
        id: number
        name: string | undefined
        thumbnail: {
            path: string
            extension: string
        }
    }
}