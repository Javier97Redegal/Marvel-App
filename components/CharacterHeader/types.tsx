export interface CharacterHeaderProps {
    id: number
}