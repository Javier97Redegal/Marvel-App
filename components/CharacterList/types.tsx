export interface CharacterListProps {
    mode: string
    query: string
}