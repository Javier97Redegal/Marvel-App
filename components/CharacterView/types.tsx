export interface CharacterViewProps {
    id: string
}