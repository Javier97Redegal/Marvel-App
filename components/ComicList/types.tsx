export interface ComicListProps {
    id: number
}