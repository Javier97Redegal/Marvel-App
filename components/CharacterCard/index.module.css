.card {
    position: relative;
}

.card img {
    display: block;
    width: 100%;
}

.card::after {
    content: '';
    display: block;
    position: absolute;
    right: 0;
    bottom: 0;
    width: 0;
    height: 0;
    border-bottom: 30px solid white;
	border-left: 30px solid transparent;
}

.image {
    width: 100%;
    aspect-ratio: 1 / 1;
    background-size: cover;
    background-position: center;
}

.title {
    display: flex;
    justify-content: space-between;
    align-items: center;
    height: 100px;
    gap: .5rem;
    padding: 2.5rem 1.5rem;
    background: black;
    border-top: 10px solid red;
}

.title a {
    font-size: 1.325rem;
    color: white;
    text-transform: uppercase;
    text-decoration: none;
    letter-spacing: 1px;
}

.title button {
    display: block;
    width: 30px;
    background: transparent;
    border: none;
    border-radius: 0;
    cursor: pointer;
    flex: 0 0 auto;
}

.title button img {
    display: block;
    width: 100%;
}