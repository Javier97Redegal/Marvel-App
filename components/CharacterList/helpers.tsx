export const getResultsMessage = (number: number): string => {
    return `Result${number === 1 ? '' : 's'}`
}