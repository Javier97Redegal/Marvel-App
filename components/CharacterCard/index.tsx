import { FC } from 'react'
import Link from 'next/link'
import { useApp } from '@/context/AppContext'
import { handleFavoriteButton } from '@/helpers'
import { GetCharactersResultType } from '@/utils/MarvelAPI/AutoGeneratedTypes/GetCharacters.types'
import styles from './index.module.css'

const CharacterCard: FC<GetCharactersResultType> = (props) => {
    const {
        id,
        name,
        thumbnail: {
            path,
            extension,
        }
    } = props
    const { favorites, addFavorite, removeFavorite } = useApp()
    const isFavorite = favorites.map(favorite => favorite.id).includes(id)

    return <>
        <div className={styles.card}>
            <div className={styles.image} style={{ backgroundImage: `url(${path}.${extension})` }}></div>
            <div className={styles.title}>
                <Link href={{ pathname: `/character/${id}` }}>{name}</Link>
                <button onClick={() => handleFavoriteButton({ character: props, isFavorite, addFavorite, removeFavorite })}>
                    {isFavorite ? <img src={'/heart-default.png'} alt='Remove from Favorites' title='Remove from Favorites' /> : <img src={'/heart-alt.png'} alt='Add to Favorites' title='Add to Favorites' />}
                </button>
            </div>
        </div>
    </>
}

export default CharacterCard