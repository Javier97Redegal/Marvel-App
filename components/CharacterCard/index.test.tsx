import '@testing-library/jest-dom'
import { render, screen, fireEvent } from '@testing-library/react'
import CharacterCard from '@/components/CharacterCard'
import { GetCharactersResultType } from '@/utils/MarvelAPI/AutoGeneratedTypes/GetCharacters.types'
import { useApp } from '@/context/AppContext'
import { handleFavoriteButton } from '@/helpers'


jest.mock('@/context/AppContext', () => ({
    useApp: jest.fn()
}))

jest.mock('@/helpers/index.tsx', () => ({
    handleFavoriteButton: jest.fn()
}))

jest.mock('next/link', () => jest.fn(({ children, href }) => <a href={href}>{children}</a>))

jest.mock('next/router', () => ({
    useRouter: jest.fn()
}))

describe('CharacterCard test', () => {
    const mockProps: GetCharactersResultType = {
        id: 12345,
        name: 'Spider-Man',
        thumbnail: {
            path: 'path/to/spiderman',
            extension: 'jpg'
        }
    }

    const mockUseApp = {
        favorites: [{ id: 12345 }],
        addFavorite: jest.fn(),
        removeFavorite: jest.fn()
    }

    beforeEach(() => {
        (useApp as jest.Mock).mockReturnValue(mockUseApp)
    })

    it('renders character details correctly', () => {
        const { asFragment } = render(<CharacterCard {...mockProps} />)

        expect(screen.getByText('Spider-Man')).toBeInTheDocument()
        expect(screen.getByRole('img', { name: 'Remove from Favorites' })).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
    })

    it('handles favorite button click', () => {
        render(<CharacterCard {...mockProps} />)

        const button = screen.getByRole('button')
        fireEvent.click(button)

        expect(handleFavoriteButton).toHaveBeenCalled()
    })

    it('displays "Add to Favorites" when not a favorite', () => {
        mockUseApp.favorites = []
        render(<CharacterCard {...mockProps} />)

        expect(screen.getByRole('img', { name: 'Add to Favorites' })).toBeInTheDocument()
    })
})
