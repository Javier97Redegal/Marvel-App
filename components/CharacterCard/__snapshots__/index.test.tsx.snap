// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterCard test renders character details correctly 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="image"
      style="background-image: url(path/to/spiderman.jpg);"
    />
    <div
      class="title"
    >
      <a
        href="[object Object]"
      >
        Spider-Man
      </a>
      <button>
        <img
          alt="Remove from Favorites"
          src="/heart-default.png"
          title="Remove from Favorites"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
